### DISCUSSION TITLE
NixOS configuration

### DISCUSSION THREAD

**Discussion by fedoranvar** [hyprwm/Hyprland#358] [General]
Good day to you all!

Primarilly, this discussion for NixOS users
and in particularily to @fufexan (because i've borrowed some ideas from his dotfiles),
but everyone is welcome

i have flake installation and have troubles with some features:

* screensharing
* launching some programs: `VirtualBox`, `vscode`, `anydesk` 

have you encountered such things NixOS users?

Thank you!

**Comment by fufexan** [comment]
I've tried all the above programs you've listed:
- VirtualBox launches fine, haven't tested but it should also work
- VSCode also launches fine, though in xwayland. Can't get it to launch with `NIXOS_OZONE_WL=1 code` in pure wayland mode, but it wouldn't work anyway https://github.com/hyprwm/Hyprland/issues/216
- Anydesk launches

Are you on the 22.05 release, or unstable?

As for the screensharing problem, I used to be able to screenshare properly on Firefox, but now it seems that it only sees XWayland applications. Possibly a misconfiguration of xdpw.

**Reply by fedoranvar** [reply]
i'm on 22.05 release, but have some packages on unstable

I've noticed you have `hyprland` in `home` attribute, what purpose it serves? why not like described in Wiki's Installation?

**Reply by fufexan** [reply]
Tbh I already had set all the options that the module defines, plus I wanted it to be inside my home config anyway, so I can use it on other OSes beside NixOS. That's all.

**Comment by viperML** [comment]
Wasn't vbox a Java app? Maybe you need `_JAVA_AWT_WM_NONREARENTING=1`

**Reply by fedoranvar** [reply]
have put in environemnt variables - didn't helped

will try to rewrite configuration

**Comment by fedoranvar** [comment]
@fufexan @viperML 
Do you know how to build on NixOS with `legacyrender`?

**Reply by fufexan** [reply]
You could try this override:
```nix
hyprland.overrideAttrs (_: {
  mesonBuildType = "legacyrenderer";
})
```

**Reply by fedoranvar** [reply]
tried 
```nix
  imports = [
    inputs.hyprland.nixosModules.default
    {
      programs.hyprland = {
        enable = true;
        package = pkgs.hyprland.overrideAttrs (_: {
          mesonBuildType = "legacyrenderer";
        });
      };
    }
  ];
```

and got 
![image](https://user-images.githubusercontent.com/52028706/181223876-1f982018-c3f9-4234-8dcf-87a451a01fd0.png)

**Reply by fufexan** [reply]
My bad, that's only available for CMake. Try `mesonFlags = ["-DLEGACY_RENDERER"];` instead of `mesonBuildType`.

**Reply by viperML** [reply]
instead,
 ```nix
hyprland.override (prev: {
  mesonFlags = prev.mesonFlags ++ ["....."];
})
```

**Reply by viperML** [reply]
Ah nevermind, the prev flags are just to **disable** XWayland

**Reply by fedoranvar** [reply]
tried 
> My bad, that's only available for CMake. Try `mesonFlags = ["-DLEGACY_RENDERER"];` instead of `mesonBuildType`.

 got 
 
![image](https://user-images.githubusercontent.com/52028706/181225498-8ed2475e-b2ff-4075-bc14-6e096b51e74c.png)

---

have tried 
` mesonFlags = [ "-DLEGACY_RENDERER=true" ];`
got
![image](https://user-images.githubusercontent.com/52028706/181225604-690614b9-49ca-4cf4-8fa6-45b70dd8aea3.png)

**Reply by fedoranvar** [reply]
got it builded by `mesonFlags = [ "-DLEGACY_RENDERER:STRING=true" ];`

**Reply by fufexan** [reply]
The `CMakeLists.txt` file doesn't lie, the [option](https://github.com/hyprwm/Hyprland/blob/main/CMakeLists.txt#L53) is there. I have no clue why it's not working, but you *could* try making an overlay from the Nixpkgs derivation (which still uses CMake for building), and add `cmakeBuildType = "legacyrenderer";` or the equivalent.

**Reply by fufexan** [reply]
Oh great.

**Reply by fedoranvar** [reply]
sadly, it didn't solved my problem with screensharing and running various programs.

-- will dig further.

by the way, is it possible to run `Hyprland` as WM of `Plasma`?

**Reply by viperML** [reply]
btw @fufexan we should change mesonFlags to be a list in nix/default.nix

**Reply by fufexan** [reply]
Will do. I also gotta add `pandoc`.

**Reply by viperML** [reply]
I don't remember if lib.optionals returned a list though, I can't check atm

**Reply by fufexan** [reply]
It does. `lib.optionals` takes a bool and a list, and `lib.optional` takes a bool and whatever else and turns it into a singleton.

**Reply by viperML** [reply]
Then I didn't say anything, mb

**Reply by fufexan** [reply]
@viperML should I add `legacyRenderer` as a nix flag?

**Reply by viperML** [reply]
seems reasonable, to improve discoverability

**Comment by fedoranvar** [comment]
update:

* i've repaired screensharing in telegram, by tweaking monitors setup in `hypr.conf`

* tried fresh Nixos installation, my nixpkgs tracking `unstable` channel, - didn't helped 

* installed similar flake on my laptop (Thinkpad T420, i7-3612QM) - everything works

* launched simultaneously `sway` and `hyprland`, and noticed, that when i'm launching `Virtual Box` in `Hyprland`, it launches in `sway` WM.

* `Plasma (Wayland)` successfully launches all programs

thoughts:
* this behaviour have something to do with bundle: `amdgpu` - `gtk`- `Hyprland`(also, months before programs was working in `Hyprland`) 
* wanna try `Archlinux` with `Hyprland` on my main machine 


suggestions? @fufexan @viperML

**Reply by fufexan** [reply]
> installed similar flake on my laptop (Thinkpad T420, i7-3612QM) - everything works

What does "similar flake" mean? The Hyprland flake is unique (and the only you should use).

> launched simultaneously `sway` and `hyprland`, and noticed, that when i'm launching `Virtual Box` in `Hyprland`, it launches in `sway` WM.

That is expected. It probably launches on `wayland-0`, which was `sway` at the time

I really don't know why it's not working for you, but I can say it's not an amdgpu problem. I have an AMD card and it works fine, Vaxerski has an AMD card and it works fine. GTK is the same story.

Not sure if Arch will help much, but you do you.

**Reply by fedoranvar** [reply]
> What does "similar flake" mean? The Hyprland flake is unique (and the only you should use).

by "similar flake" i meant my flake-dotfiles

> That is expected. It probably launches on wayland-0, which was sway at the time

i've tried removing `sway` and `plasma`, but still nothing
right now `WAYLAND_DISPLAY` echoes `wayland-0` under `Hyprland`
and my `XDG_CURRENT_DESKTOP` is `hyprland`

**Reply by fedoranvar** [reply]
![image](https://user-images.githubusercontent.com/52028706/181872847-9a33a558-49aa-4f6c-a984-58ec0419dddf.png)

as screensharing now working in telegram, i'm able to screenshare it, but there is no `Virtualbox` client in `hyprctl clients`

**Reply by fedoranvar** [reply]
also, @fufexan 

i've noticed you're using `swayidle`

how to make something like `dpms off` (black screens) in `Hyprland`?

**Reply by fufexan** [reply]
I'm using it for sway itself, haven't got around to also making it work for hyprland.

The dpms off equivalent on hyprland is `hyprctl keyword monitor DP-1,off`.

**Reply by vaxerski** [reply]
incorrect. First of all, it's `disabled`, second, `disabled` will not do what fedoranvar thinks I am afraid, fedon thinks about turning screens off as a screensaver, `disabled` will turn off the monitor completely, making it unmanaged by hyprland and throwing all workspaces to a different monitor. Having all monitors off will exit hyprland. (As there is no monitor left)

**Reply by fufexan** [reply]
Then we don't have DPMS I'm afraid. sad

**Reply by vaxerski** [reply]
we dont

**Comment by fedoranvar** [comment]
i'm closing this discussion, as i've moved to thinkpad , and everything works there.

But for my case, i think it have something to do with xwayland side - besides don't know.

Thanks to everyone!

**Reply by fedoranvar** [reply]
UPD. 

I've returned to my 'troubled' system, and everything works.

Don't know why.
